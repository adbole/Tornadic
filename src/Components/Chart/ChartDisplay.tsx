import { AreaChart, BarChart, LineChart } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";

import { ChartViews } from ".";


type ChartDisplayProps = {
    children: React.ReactNode;
    property: ChartViews;
} & CategoricalChartProps;

//Provides the proper chart to display based on the HourlyProperty passed
export default function ChartDisplay(props: ChartDisplayProps) {
    const { children, property, ...excess } = props;

    let ChosenChart = BarChart;

    switch (property) {
        case "precipitation":
            ChosenChart = BarChart;
            break;
        case "temperature_2m":
        case "uv_index":
        case "us_aqi":
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return <ChosenChart {...excess}>{children}</ChosenChart>;
}
