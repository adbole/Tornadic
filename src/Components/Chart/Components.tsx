import { AreaChart, BarChart, LineChart, TooltipProps } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";
import { NameType } from "recharts/types/component/DefaultTooltipContent";

import { UV_MAX_VALUES } from "ts/Constants";

import { ChartViews, DataPoint } from ".";

function getUnit(data: DataPoint, defaultUnit: React.ReactNode) {
    const { property, primaryKey } = data;

    if(property === ChartViews.UV_Index) {
        if (primaryKey <= UV_MAX_VALUES.LOW) return " Low"; 
        else if (primaryKey <= UV_MAX_VALUES.MODERATE) return " Moderate";
        else if (primaryKey <= UV_MAX_VALUES.HIGH) return " High";
        else if (primaryKey <= UV_MAX_VALUES.VERY_HIGH) return " Very High";
        else return " Extreme"; 
    }

    return defaultUnit;
}
  
//Customized tooltip to display primary and secondary data with labels if applicable
export const CustomTooltip = ({ active, payload }: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const fixDecimal = (property: ChartViews, value: number): string | number => 
        (property === ChartViews.Precipitation || property === ChartViews.Pressure) ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    let unit = getUnit(data, payload[0].unit);
    const label = (data.property === ChartViews.Temperature && "Feels: ") || (data.property === ChartViews.Windspeed && "Gust: ");


    return (
        <div className="chart-tooltip">
            <h1>{fixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{fixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

type ChartDisplayProps = {
    children: React.ReactNode,
    property: ChartViews
} & CategoricalChartProps

//Provides the proper chart to display based on the HourlyProperty passed
export const ChartDisplay = (props: ChartDisplayProps) => {
    const { children, property, ...excess } = props;

    let ChosenChart = BarChart;

    switch(property) {
        case ChartViews.Precipitation:
            ChosenChart = BarChart;
            break;
        case ChartViews.Temperature:
        case ChartViews.UV_Index:
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return (
        <ChosenChart {...excess}>
            {children}
        </ChosenChart>
    );
};