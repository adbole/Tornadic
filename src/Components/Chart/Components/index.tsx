export { default as Area } from "./Area"
export { default as Axes } from "./Axes"
export { default as Bar } from "./Bar"
export { default as Line } from "./Line"
export { default as NowReference } from "./NowReference"
export { default as Tooltip, useTooltip } from "./Tooltip"