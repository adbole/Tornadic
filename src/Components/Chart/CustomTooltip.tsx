import { TooltipProps } from "recharts";
import { NameType } from "recharts/types/component/DefaultTooltipContent";

import { get_aq, get_uv } from "ts/Helpers";

import { ChartViews, DataPoint } from ".";


function getUnit(data: DataPoint, defaultUnit: React.ReactNode) {
    const { property, primaryKey } = data;

    if(property === "uv_index") {
        return " " + get_uv(primaryKey);
    }
    else if(property === "us_aqi") {
        return " " + get_aq(primaryKey);
    }

    return defaultUnit;
}

//Customized tooltip to display primary and secondary data with labels if applicable
const CustomTooltip = ({ active, payload }: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const fixDecimal = (property: ChartViews, value: number): string | number => 
        (property === "precipitation" || property === "surface_pressure") ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    const unit = getUnit(data, payload[0].unit);
    const label = 
        (data.property === "temperature_2m" && "Feels: ") || 
        (data.property === "windspeed_10m" && "Gust: ");

    return (
        <div className="chart-tooltip">
            <h1>{fixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{fixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

export default CustomTooltip;