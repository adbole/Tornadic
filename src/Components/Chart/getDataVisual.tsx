import { Area, Bar, Line } from "recharts";

import { get_aq, get_uv, nameof, toHSL } from "ts/Helpers";

import { ChartViews, DataPoint } from ".";


export default function getDataVisual(unit: string, view: ChartViews, dataPoints: DataPoint[]) {
    switch(view) {
        case "precipitation":
            return <Bar dataKey={nameof<DataPoint>("primaryKey")} fill={"#0078ef"} unit={unit}/>;
        case "temperature_2m": {
            const dataValues = dataPoints.flatMap(point => [point.primaryKey, (point.secondaryKey as number) ?? 0]);
            const min = Math.min(...dataValues);
            const max = Math.max(...dataValues);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="0%" stopColor={toHSL(Math.min(120, max))}/>
                            <stop offset="100%" stopColor={toHSL(Math.max(0, min))}/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={0.75} fill="url(#tempGradient)" unit={unit}/>
                    <Area type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#fff" fillOpacity={0}/>
                </>
            );
        }
        case "uv_index": {
            const maxUV = Math.max(...dataPoints.map(point => point.primaryKey));
            const maxLevel = get_uv(maxUV);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            {maxLevel === "Extreme" && <stop offset="0%" stopColor="#FF00D6"/>}
                            {maxLevel === "Very High" && <stop offset="25%" stopColor="#FF2204"/>}
                            {maxLevel === "High" && <stop offset="50%" stopColor="#FF9431"/>}
                            {maxLevel === "Moderate" && <stop offset="75%" stopColor="#FFF501"/>}
                            <stop offset="100%" stopColor="#00FF66"/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={0.75} fill="url(#tempGradient)"/>
                </>
            );
        }
        case "us_aqi": {
            const maxAQ = Math.max(...dataPoints.map(point => point.primaryKey));
            const maxLevel = get_aq(maxAQ);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            {maxLevel === "Hazardous" && <stop offset="0%" stopColor="#6D0000"/>}
                            {maxLevel === "Very Unhealthy" && <stop offset="20%" stopColor="#8400FF"/>}
                            {maxLevel === "Unhealthy" && <stop offset="40%" stopColor="#FF2204"/>}
                            {maxLevel === "Unhealthy*" && <stop offset="60%" stopColor="#FF9431"/>}
                            {maxLevel === "Moderate" && <stop offset="80%" stopColor="#FFF501"/>}
                            <stop offset="100%" stopColor="#00FF66"/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={0.75} fill="url(#tempGradient)"/>
                </>
            );
        }
        default:
            return (
                <>
                    <Line type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#2668f7" unit={unit}/>
                    {dataPoints[0].secondaryKey != null && <Line type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#2eff7d"/>}
                </>
            );
    }
}