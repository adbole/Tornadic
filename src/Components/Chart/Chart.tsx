import { ResponsiveContainer, Line, CartesianGrid, XAxis, YAxis, Tooltip, Bar, Area, ReferenceLine } from "recharts";
import { CustomTooltip, PressureTick, ChartDisplay } from "./Chart.Components";

import { Forecast, useWeather } from "../Contexes/Weather";
import React from "react";
import { TimeConverter, nameof } from "../../ts/Helpers";
import { Modal, ModalContent, ModalTitle } from "../Contexes/ModalContext";

//Properties denoted here represent values supported for displaying by a Chart
export enum HourlyProperties {
    Temperature = "temperature_2m",
    Humidity = "relativehumidity_2m",
    Precipitation = "precipitation",
    Dewpoint = "dewpoint_2m",
    Visibility = "visibility",
    Windspeed = "windspeed_10m",
    Pressure = "surface_pressure"
}

export type DataPoint = {
    property: HourlyProperties,
    name: string,
    primaryKey: number,
    secondaryKey: number | null
}

const ToHSL = (x: number) => `hsl(${250 * ((120-x)/120)}deg, 100%, 50%)`;

//Gets the data for the given day and property
function GetData(forecastData: Forecast, property: HourlyProperties, day: number) {
    let data: DataPoint[] = [];

    for(let i = 24 * (day); i < 24 * (day + 1); ++i) {
        data.push({
            property: property,
            name: TimeConverter.GetTimeFormatted(forecastData.hourly.time[i], TimeConverter.TimeFormat.Hour),
            primaryKey: forecastData.hourly[property][i],
            secondaryKey: GetSecondaryKey(i)
        });
    }

    return data;

    //Some properties have secondary values
    function GetSecondaryKey(i: number) {
        switch(property) {
            case HourlyProperties.Temperature:
                return forecastData.hourly.apparent_temperature[i];
            case HourlyProperties.Windspeed:
                return forecastData.hourly.windgusts_10m[i];
            default:
                return null;
        }
    }
}

function GetMinMax([min, max]: [number, number], property: HourlyProperties): [number, number] {
    if(property === HourlyProperties.Pressure) {
        min -= 0.1;
        max += 0.1;
    }
    else if(property === HourlyProperties.Precipitation) {
        min = 0;
        max = Math.max(0.5, max);
    }
    else if (property === HourlyProperties.Humidity) {
        min = 0;
        max = 100;
    }
    else {
        min = Math.floor((min * 0.95) / 10) * 10;
        max = Math.ceil((max * 1.10) / 10) * 10;
    }

    return [min, max];
}

function GetDataVisual(forecastData: Forecast, property: HourlyProperties, data: DataPoint[]) {
    switch(property) {
        case HourlyProperties.Precipitation:
            return <Bar dataKey={nameof<DataPoint>("primaryKey")} fill={"#0078ef"} />;
        case HourlyProperties.Temperature:
            const dataValues = data.flatMap(point => [point.primaryKey, (point.secondaryKey as number) ?? 0]);
            const minMax = GetMinMax([Math.min(...dataValues), Math.max(...dataValues)], property);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="0%" stopColor={ToHSL(Math.min(120, minMax[1]))} stopOpacity="0.5"/>
                            <stop offset="100%" stopColor={ToHSL(Math.max(0, minMax[0]))} stopOpacity="0.5"/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={1} fill="url(#tempGradient)" unit={forecastData.hourly_units[property]}/>
                    <Area type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#fff" fillOpacity={0}/>
                </>
            );
        default:
            return (
                <>
                    <Line type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#2668f7" unit={forecastData.hourly_units[property]}/>
                    {data[0].secondaryKey != null && <Line type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#2eff7d"/>}
                </>
            );
    }
}

const Chart = ({showProperty, showDay = 0}: {showProperty: HourlyProperties, showDay?: number}) => {
    const forecastData = useWeather().forecast;
    const [property, setProperty] = React.useState(showProperty);
    const [day, setDay] = React.useState(showDay);

    const chartData = React.useMemo(() => GetData(forecastData, property, day), [forecastData, property, day]);

    const timeRef = React.useRef<HTMLSpanElement>(null);
    const setTimeText = React.useCallback((s: string) => {
        if(!timeRef.current) return;

        timeRef.current.innerText = s;
    }, []);

    const onMouseMove = React.useCallback((e: any) => e.activeLabel ? setTimeText(", " + e.activeLabel) : setTimeText(""), [setTimeText]);
    const onMouseLeave = React.useCallback(() => setTimeText(""), [setTimeText]);

    return (
       <Modal id="chart">
            <ModalTitle>
                <select className="clear" title="Current Chart" onChange={(e) => setProperty(e.currentTarget.value as HourlyProperties)} value={property}>
                    {
                        Object.keys(HourlyProperties).map(key => (
                            <option key={key} value={HourlyProperties[key as keyof typeof HourlyProperties]}>{key}</option>
                        ))
                    }
                </select>
            </ModalTitle>
            <ModalContent>
                <div className="controls">
                    {
                        forecastData.hourly.time.filter((_, i) =>  i % 24 === 0).map((time, i) => (
                            <div key={i} className="toggle-button" onClick={() => setDay(i)}>
                                <input type="radio" name="chart-radio" id={i.toString()} defaultChecked={i === day}/>
                                <label htmlFor={i.toString()}>{TimeConverter.GetTimeFormatted(time, TimeConverter.TimeFormat.Weekday)}</label>
                            </div>
                        ))
                    }
                </div>

                <p><span>{TimeConverter.GetTimeFormatted(forecastData.hourly.time[day * 24], TimeConverter.TimeFormat.Date)}</span><span ref={timeRef}></span></p>

                <hr></hr>

                <ResponsiveContainer width={"100%"} height="100%">
                    <ChartDisplay property={property} data={chartData} margin={{top: 0, left: 0, right: 0, bottom: 0}} onMouseMove={onMouseMove} onMouseLeave={onMouseLeave}>
                        {GetDataVisual(forecastData, property, chartData)}
                        <CartesianGrid stroke="#ffffff19"/>
                        <XAxis dataKey="name" interval={5} textAnchor="start"/>
                        {
                            property === HourlyProperties.Pressure 
                            ? <YAxis domain={([dataMin, dataMax]) => GetMinMax([dataMin, dataMax], property)} tick={<PressureTick unit={forecastData.hourly_units[property]}/>}/>
                            : <YAxis domain={([dataMin, dataMax]) => GetMinMax([dataMin, dataMax], property)} unit={forecastData.hourly_units[property]} />
                        }
                        
                        {/* @ts-ignore */}
                        <Tooltip wrapperStyle={{ outline: "none" }} position={{x: 'auto', y: 10}} content={<CustomTooltip/>}/>
                        {day === 0 && <ReferenceLine x={TimeConverter.GetTimeFormatted(forecastData.hourly.time[forecastData.nowIndex], TimeConverter.TimeFormat.Hour)}/>}
                    </ChartDisplay>
                </ResponsiveContainer>
            </ModalContent>
       </Modal>
    );
};

export default React.memo(Chart);