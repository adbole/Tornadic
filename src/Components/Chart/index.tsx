import { ResponsiveContainer, Line, CartesianGrid, XAxis, YAxis, Tooltip, Bar, Area, ReferenceLine } from "recharts";
import { CustomTooltip, ChartDisplay } from "./Components";

import { Forecast, useWeather } from "Components/Contexts/Weather";
import React from "react";
import { TimeConverter, nameof } from "ts/Helpers";
import { Modal, ModalContent, ModalTitle } from "Components/Contexts/ModalContext";

//Properties denoted here represent values supported for displaying by a Chart
export enum HourlyProperties {
    Temperature = "temperature_2m",
    Humidity = "relativehumidity_2m",
    Precipitation = "precipitation",
    Dewpoint = "dewpoint_2m",
    Visibility = "visibility",
    Windspeed = "windspeed_10m",
    Pressure = "surface_pressure"
}

export type DataPoint = {
    property: HourlyProperties,
    name: string,
    primaryKey: number,
    secondaryKey: number | null
}

const ToHSL = (x: number) => `hsl(${250 * ((120-x)/120)}deg, 100%, 50%)`;

//Gets the data for the given day and property
function GetData(forecastData: Forecast, property: HourlyProperties, day: number) {
    let data: DataPoint[] = [];

    for(let i = 24 * (day); i < 24 * (day + 1); ++i) {
        data.push({
            property: property,
            name: TimeConverter.GetTimeFormatted(forecastData.hourly.time[i], TimeConverter.TimeFormat.Hour),
            primaryKey: forecastData.hourly[property][i],
            secondaryKey: GetSecondaryKey(i)
        });
    }

    return data;

    //Some properties have secondary values
    function GetSecondaryKey(i: number) {
        switch(property) {
            case HourlyProperties.Temperature:
                return forecastData.hourly.apparent_temperature[i];
            case HourlyProperties.Windspeed:
                return forecastData.hourly.windgusts_10m[i];
            default:
                return null;
        }
    }
}

function GetMinMax([min, max]: [number, number], property: HourlyProperties): [number, number] {
    if(property === HourlyProperties.Pressure) {
        min -= 0.3;
        max += 0.3;
    }
    else if(property === HourlyProperties.Precipitation) {
        min = 0;
        max = Math.max(0.5, max);
    }
    else if (property === HourlyProperties.Humidity) {
        min = 0;
        max = 100;
    }
    else {
        min = Math.floor((min * 0.95) / 10) * 10;
        max = Math.ceil((max * 1.10) / 10) * 10;
    }

    return [min, max];
}

function GetDataVisual(forecastData: Forecast, property: HourlyProperties, data: DataPoint[]) {
    switch(property) {
        case HourlyProperties.Precipitation:
            return <Bar dataKey={nameof<DataPoint>("primaryKey")} fill={"#0078ef"} />;
        case HourlyProperties.Temperature:
            const dataValues = data.flatMap(point => [point.primaryKey, (point.secondaryKey as number) ?? 0]);
            const minMax = GetMinMax([Math.min(...dataValues), Math.max(...dataValues)], property);

            return (
                <>
                    <defs>
                        <linearGradient id="tempGradient" x1="0" y1="0" x2="0" y2="1">
                            <stop offset="0%" stopColor={ToHSL(Math.min(120, minMax[1]))} stopOpacity="0.5"/>
                            <stop offset="100%" stopColor={ToHSL(Math.max(0, minMax[0]))} stopOpacity="0.5"/>
                        </linearGradient>
                    </defs>
                    <Area type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#ffffff00" fillOpacity={1} fill="url(#tempGradient)" unit={forecastData.hourly_units[property]}/>
                    <Area type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#fff" fillOpacity={0}/>
                </>
            );
        default:
            return (
                <>
                    <Line type="monotone" dataKey={nameof<DataPoint>("primaryKey")} stroke="#2668f7" unit={forecastData.hourly_units[property]}/>
                    {data[0].secondaryKey != null && <Line type="monotone" dataKey={nameof<DataPoint>("secondaryKey")} stroke="#2eff7d"/>}
                </>
            );
    }
}

const Chart = ({showProperty, showDay = 0}: {showProperty: HourlyProperties, showDay?: number}) => {
    const forecastData = useWeather().forecast;
    const [property, setProperty] = React.useState(showProperty);
    const [day, setDay] = React.useState(showDay);

    const chartData = React.useMemo(() => GetData(forecastData, property, day), [forecastData, property, day]);

    const timeRef = React.useRef<HTMLSpanElement>(null);
    const selectRef = React.useRef<HTMLSelectElement>(null);

    //Autosize the select element for style points
    React.useEffect(() => {
        if(!selectRef.current) return;

        const canvasContext = document.createElement('canvas').getContext('2d')!;
        //16px is default font-size and select is within a h1 getting 2rem font-size, therefore 32px is used here.
        canvasContext.font = '32px system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Open Sans", "Helvetica Neue", sans-serif'

        //Incase textContent is null Temperature is default since its the largest option
        const width = canvasContext.measureText(selectRef.current.children[selectRef.current.selectedIndex].textContent ?? "Temperature").width

        selectRef.current.style.width = Math.round(width) + 20 + "px";
    }, [property, selectRef])

    const setTimeText = React.useCallback((s: string) => {
        if(!timeRef.current) return;

        timeRef.current.innerText = s;
    }, []);

    const onMouseMove = React.useCallback((e: any) => e.activeLabel ? setTimeText(", " + e.activeLabel) : setTimeText(""), [setTimeText]);
    const onMouseLeave = React.useCallback(() => setTimeText(""), [setTimeText]);

    return (
       <Modal id="chart">
            <ModalTitle>
                <select ref={selectRef} className="clear" title="Current Chart" onChange={(e) => setProperty(e.currentTarget.value as HourlyProperties)} value={property}>
                    {
                        Object.keys(HourlyProperties).map(key => (
                            <option key={key} value={HourlyProperties[key as keyof typeof HourlyProperties]}>{key}</option>
                        ))
                    }
                </select>
            </ModalTitle>
            <ModalContent>
                <div className="controls">
                    {
                        forecastData.hourly.time.filter((_, i) =>  i % 24 === 0).map((time, i) => (
                            <div key={i} className="toggle-button" onClick={() => setDay(i)}>
                                <input type="radio" name="chart-radio" id={i.toString()} defaultChecked={i === day}/>
                                <label htmlFor={i.toString()}>{TimeConverter.GetTimeFormatted(time, TimeConverter.TimeFormat.Weekday)}</label>
                            </div>
                        ))
                    }
                </div>

                <p><span>{TimeConverter.GetTimeFormatted(forecastData.hourly.time[day * 24], TimeConverter.TimeFormat.Date)}</span><span ref={timeRef}></span></p>

                <hr></hr>

                <ResponsiveContainer width={"100%"} height="100%">
                    <ChartDisplay property={property} data={chartData} margin={{top: 0, left: 0, right: 0, bottom: 0}} onMouseMove={onMouseMove} onMouseLeave={onMouseLeave}>
                        {GetDataVisual(forecastData, property, chartData)}
                        <CartesianGrid stroke="#ffffff19"/>
                        <XAxis dataKey="name" interval={5} textAnchor="start"/>
                        {
                            property == HourlyProperties.Temperature || property === HourlyProperties.Dewpoint || property === HourlyProperties.Humidity
                            ? <YAxis width={50} domain={([dataMin, dataMax]) => GetMinMax([dataMin, dataMax], property)} unit={forecastData.hourly_units[property]}/>
                            : <YAxis width={50} domain={([dataMin, dataMax]) => GetMinMax([dataMin, dataMax], property)} tickFormatter={(value: number) => (Math.round(value * 10) / 10).toString()}/>
                        }
                        
                        {/* @ts-ignore */}
                        <Tooltip wrapperStyle={{ outline: "none" }} position={{x: 'auto', y: 10}} content={<CustomTooltip/>}/>
                        {day === 0 && <ReferenceLine x={TimeConverter.GetTimeFormatted(forecastData.hourly.time[forecastData.nowIndex], TimeConverter.TimeFormat.Hour)}/>}
                    </ChartDisplay>
                </ResponsiveContainer>
            </ModalContent>
       </Modal>
    );
};

export default React.memo(Chart);