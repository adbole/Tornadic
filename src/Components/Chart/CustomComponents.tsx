import { TooltipProps, BarChart, AreaChart, LineChart } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";
import { NameType } from "recharts/types/component/DefaultTooltipContent";
import { DataPoint, HourlyProperties } from "./Chart";

//Airpressure needs a special tick in order to display properly
export const PressureTick = ({x, y, payload, unit}: {x: number, y:number, payload: {value: number}, unit: string} & any) => (
    <g transform={`translate(${x},${y})`}>
        <text x={0} y={0} dy={5} fill="#666">
        <tspan textAnchor="end" x="0">{payload.value.toFixed(2)}</tspan>
        <tspan textAnchor="end" x="0" dy="20">{unit}</tspan>
        </text>
    </g>
);
  
//Customized tooltip to display primary and secondary data with labels if applicable
export const CustomTooltip = ({active, payload}: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const FixDecimal = (property: HourlyProperties, value: number): string | number => 
        (property === HourlyProperties.Precipitation || property === HourlyProperties.Pressure) ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    const unit = payload[0].unit;
    const label = (data.property === HourlyProperties.Temperature && "Feels: ") || (data.property === HourlyProperties.Windspeed && "Gust: ");

    return (
        <div className="chart-tooltip">
            <h1>{FixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{FixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

type ChartDisplayProps = {
    children: React.ReactNode,
    property: HourlyProperties
} & CategoricalChartProps

//Provides the proper chart to display based on the HourlyProperty passed
export const ChartDisplay = (props: ChartDisplayProps) => {
    const {children, property, ...excess} = props;

    let ChosenChart = BarChart;

    switch(property) {
        case HourlyProperties.Precipitation:
            ChosenChart = BarChart;
            break;
        case HourlyProperties.Temperature:
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return (
        <ChosenChart {...excess}>
            {children}
        </ChosenChart>
    );
};