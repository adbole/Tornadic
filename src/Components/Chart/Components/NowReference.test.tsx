import { forecast, useWeather } from "@test-mocks";
import { mockDate } from "@test-utils";

import { render } from "@testing-library/react";

import Chart from "Components/Chart";

import { NowReference } from ".";


mockDate();

vi.mock("Contexts/WeatherContext", () => ({ useWeather }));

const dataPoints = forecast().hourly.time.slice(0, 24).map((iso, i) => ({ 
    x: new Date(iso), 
    y: [i] 
}));

function Wrapper({ children }: { children: React.ReactNode }) {
    return (
        <Chart dataPoints={dataPoints} type="linear">
            {children}
        </Chart>
    );
}

test("When shown, draws a rect and line to the current time", () => {
    const { container } = render(
        <Wrapper>
            <NowReference isShown />
        </Wrapper>
    );

    expect(container).toMatchSnapshot();
});

test("When shown, gives a offset for bandWidth", () => {
    const { container } = render(
        <Chart dataPoints={dataPoints} type="band">
            <NowReference isShown />
        </Chart>
    );

    expect(container).toMatchSnapshot();
});

test("When not shown, returns null", () => {
    const { container } = render(
        <Wrapper>
            <NowReference isShown={false} />
        </Wrapper>
    );

    expect(container.querySelector("svg")).toBeEmptyDOMElement();
});
