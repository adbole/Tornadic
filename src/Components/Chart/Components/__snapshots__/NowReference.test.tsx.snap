// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`When shown, draws a rect and line to the current time 1`] = `
.emotion-0 {
  position: relative;
  border: 1px solid #ffffff19;
  border-radius: var(--border-radius);
  overflow: hidden;
  width: 100%;
  height: 100%;
  display: block;
}

<div>
  <svg
    class="emotion-0 emotion-1"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 0 0"
  >
    <g>
      <rect
        fill="#000"
        fill-opacity="0.25"
        height="-140"
        width="-45.65217391304348"
        x="40"
        y="110"
      />
      <line
        stroke="#fff"
        stroke-width="1"
        x1="-5.652173913043476"
        x2="-5.652173913043476"
        y1="110"
        y2="-30"
      />
    </g>
  </svg>
</div>
`;

exports[`When shown, gives a offset for bandWidth 1`] = `
.emotion-0 {
  position: relative;
  border: 1px solid #ffffff19;
  border-radius: var(--border-radius);
  overflow: hidden;
  width: 100%;
  height: 100%;
  display: block;
}

<div>
  <svg
    class="emotion-0 emotion-1"
    preserveAspectRatio="xMidYMid meet"
    viewBox="0 0 0 0"
  >
    <g>
      <rect
        fill="#000"
        fill-opacity="0.25"
        height="-140"
        width="-44.70954356846473"
        x="40"
        y="110"
      />
      <line
        stroke="#fff"
        stroke-width="1"
        x1="-4.70954356846473"
        x2="-4.70954356846473"
        y1="110"
        y2="-30"
      />
    </g>
  </svg>
</div>
`;
